src/null-safety/unsound-null-safety.md

```diff
@@ -0,0 +1,178 @@
+---
+
+An app can use some libraries that
+are [null safe][] and some that aren't.
+Such an app — called a **mixed-version program** —
+executes with **unsound null safety**.
+
+[null safe]: /null-safety
+
+Mixing [language versions][] lets you migrate to null safety at your own pace.
+It also frees package maintainers to migrate their code,
+with the knowledge that even legacy users can get new
+bug fixes and other improvements.
+However, mixed-version programs don't get all the advantages
+that null safety can bring.
+
+[language versions]: /guides/language/evolution#language-versioning
+
+This page describes the differences between sound and unsound null safety,
+to help you decide whether to migrate before
+all your dependencies support null safety.
+After the conceptual discussion are instructions for migrating incrementally,
+followed by details on analyzing, testing, and running mixed-version programs.
+
+{{ site.alert.note }}
+  We recommend that, if possible, you wait for dependencies to migrate
+  before you migrate your app.
+  For details, see the [migration guide][].
+{{ site.alert.end }}
+
+[migration guide]: /null-safety/migration-guide
+
+
+## Sound and unsound null safety
+
+Dart provides sound null safety through a combination of
+static and runtime checks.
+Each Dart library that opts in to null safety gets
+all the _static_ checks, with stricter compile-time errors.
+This is true even in a mixed-version program that contains
+null-unsafe libraries.
+You start getting these benefits
+as soon as you start migrating some of your code to null safety.
+
+However, a mixed-version program can't have the
+_runtime_ soundness guarantees that a fully null-safe app has.
+It's possible for `null` to leak out of the null-unsafe libraries
+into the null-safe code, because
+preventing that would break the existing behavior of the unmigrated code.
+
+To maintain runtime compatibility with legacy libraries
+while offering soundness to completely null-safe programs,
+Dart tools support two modes:
+
+* When you run a mixed-version program,
+  it runs with **unsound null safety**.
+  It's possible for `null` reference errors to occur at runtime,
+  but only because a `null` or nullable type escaped from
+  some null-unsafe library and got into your null-safe code.
+
+* When your program is fully migrated and _all_ libraries are null safe,
+  then your program automatically runs with **sound null safety**.
+  That mode gives you the full null safety experience with
+  all of the guarantees and compiler optimizations that soundness enables.
+
+Sound null safety is what you want if possible.
+Dart tools automatically run your program in sound mode if
+the main entrypoint library of your program has opted into null safety.
+If you import a null-unsafe library,
+the tools print a warning to let you know that
+they can only [run with unsound null safety](#analyzing-and-testing).
+
+
+## Migrating incrementally
```

Comment:
Is there a reason we document incremental manual migration here, but not using the migration tool?

Reply:
I'm not sure... I think we wanted to add better support for incremental migration to the tool, so maybe that's why.

Reply:
Yeah, the tool doesn't really support this yet, except via a really kludgy workflow.  I think it's better to follow up here later if/when the tool gets support for this.  cc @stereotype441

Reply:
FYI, incremental migration work is being tracked at https://github.com/dart-lang/sdk/issues/44124